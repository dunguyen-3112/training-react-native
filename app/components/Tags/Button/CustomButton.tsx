import { Image, StyleSheet, ImageSourcePropType } from 'react-native';
import React from 'react';
import { COLOR, COLORS, Font } from '@constants';
import CustomText from '@components/CustomText';
import { TouchableOpacity } from 'react-native-gesture-handler';

interface ButtonStyle {
  color?: COLOR;
  label: string;
  icon?: ImageSourcePropType;
  labelFont?: Font;
  labelColor?: COLOR;
  marginTop?: number;
}

const CustomButton = ({
  color = 'green',
  label,
  marginTop = 0,
  icon,
  labelFont = { fontSize: 12, fontWeight: '400' },
  labelColor = 'white',
}: ButtonStyle) => {
  const _color = color === 'green' ? COLORS.GREEN_COLOR : COLORS.ORANGE_COLOR;
  return (
    <TouchableOpacity
      style={[styles.button, { backgroundColor: _color, marginTop: marginTop }]}
    >
      <CustomText
        text={label}
        size={labelFont.fontSize}
        weight={labelFont.fontWeight}
        color={labelColor}
      />
      <Image source={icon} style={styles.icon} />
    </TouchableOpacity>
  );
};

export default CustomButton;

const styles = StyleSheet.create({
  button: {
    flexDirection: 'row',
    color: '',
    paddingVertical: 6,
    paddingHorizontal: 19,
    borderRadius: 8,
    justifyContent: 'center',
    alignItems: 'center',
    alignSelf: 'flex-start',
  },
  icon: {
    marginLeft: 8,
  },
});
