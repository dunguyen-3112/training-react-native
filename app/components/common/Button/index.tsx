import React, { memo, ReactNode, useMemo } from 'react';
import { Pressable, StyleSheet, Text, ViewStyle } from 'react-native';

import { COLORS, FONT_SIZES, FONT_WEIGHTS } from '@constants';
import { getFont } from '@helpers';

const BASE_STYLE = {
  fontSize: FONT_SIZES['ms-3'],
  fontWeight: FONT_WEIGHTS.m,
  color: COLORS.GRAY,
};

const BUTTON_TYPES = {
  primary: {
    backgroundColor: COLORS.PRIMARY,
    color: COLORS.WHITE,
    fontSize: FONT_SIZES['xxl-0'],
    fontWeight: FONT_WEIGHTS.b,
  },
  green: {
    backgroundColor: COLORS.GREEN,
    color: COLORS.WHITE,
    fontSize: FONT_SIZES.ms,
    fontWeight: FONT_WEIGHTS.b,
  },
  default: {
    backgroundColor: COLORS.LIGHT_GRAY,
    ...BASE_STYLE,
  },
  secondary: {
    backgroundColor: COLORS.SECONDARY,
    color: COLORS.WHITE,
    fontSize: FONT_SIZES['xxl-0'],
    fontWeight: FONT_WEIGHTS.b,
  },
  active: {
    backgroundColor: COLORS.LIGHT_GREEN,
    ...BASE_STYLE,
    color: COLORS.PRIMARY,
  },
};

export type BUTTON_TYPE =
  | 'primary'
  | 'secondary'
  | 'active'
  | 'default'
  | 'green';

export interface ButtonProps {
  label?: string;
  width?: number | string;
  type?: BUTTON_TYPE;
  height?: number;
  children?: ReactNode;
  borderRadius?: number;
  onPress?: () => void;
  marginTop?: number;
  paddingVertical?: number;
  paddingHorizontal?: number;
  customStyle?: ViewStyle;
}

const Button = ({
  label,
  type = 'default',
  children,
  borderRadius,
  paddingVertical,
  paddingHorizontal,
  customStyle,
  onPress,
}: ButtonProps) => {
  const TB = BUTTON_TYPES[type];
  const { fontWeight } = TB;

  const _fontWeight = useMemo(() => getFont(fontWeight), [fontWeight]);

  return (
    <Pressable
      onPress={onPress}
      style={[
        styles.button,
        {
          ...TB,
          paddingVertical,
          paddingHorizontal,
          borderRadius,
          ...customStyle,
        },
      ]}
    >
      {children ? (
        children
      ) : (
        <Text style={{ ...TB, fontWeight: _fontWeight }}>{label}</Text>
      )}
    </Pressable>
  );
};

export default memo(Button);

const styles = StyleSheet.create({
  button: {
    flexDirection: 'row',
    borderRadius: 8,
    justifyContent: 'center',
    alignItems: 'center',
    alignSelf: 'flex-start',
  },
});
