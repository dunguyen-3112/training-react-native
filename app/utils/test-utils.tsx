import { render, RenderOptions } from '@testing-library/react-native';
import React, { ReactElement } from 'react';
import { NavigationContainer } from '@react-navigation/native';

const AllTheProviders = ({ children }: { children: ReactElement }) => {
  return <NavigationContainer>{children}</NavigationContainer>;
};

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react-native';

export { customRender as render };
