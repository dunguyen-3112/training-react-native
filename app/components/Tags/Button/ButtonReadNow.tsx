import { COLOR } from '@constants';
import React from 'react';
import CustomButton from './CustomButton';

const ButtonReadNow = ({
  marginTop = 0,
  color,
}: {
  marginTop?: number;
  color?: COLOR;
}) => {
  return (
    <CustomButton
      label="Read now"
      icon={require('@assets/icons/Arrow.png')}
      marginTop={marginTop}
      color={color}
    />
  );
};

export default ButtonReadNow;
