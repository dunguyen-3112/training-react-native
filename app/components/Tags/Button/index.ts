import ButtonReadNow from './ButtonReadNow';
import CustomButton from './CustomButton';

export { ButtonReadNow, CustomButton };
