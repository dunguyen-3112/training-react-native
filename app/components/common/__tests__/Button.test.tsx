import React from 'react';
import Button, { ButtonProps } from '../Button';
import { render, screen } from '@utils/test-utils';

const setup = (props: ButtonProps = {}) => {
  render(<Button {...props} />);
};

describe('Button', () => {
  it('should render default props', () => {
    setup({ label: 'Button' });
    expect(screen.getByText('Button')).toBeTruthy();
  });

  it('should render type primary', () => {
    setup({ label: 'Button', type: 'primary' });
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it('should render type secondary', () => {
    setup({ label: 'Button', type: 'secondary' });
    expect(screen.toJSON()).toMatchSnapshot();
  });
});
