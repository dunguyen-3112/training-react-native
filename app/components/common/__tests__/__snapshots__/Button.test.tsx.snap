// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button has a default 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "alignSelf": "flex-start",
        "borderRadius": 8,
        "flexDirection": "row",
        "justifyContent": "center",
      },
      {
        "backgroundColor": "#91C788",
        "borderRadius": undefined,
        "paddingHorizontal": undefined,
        "paddingVertical": undefined,
      },
    ]
  }
>
  <Text
    onPress={[Function]}
    style={
      [
        {
          "borderColor": "#5C5C5C",
          "color": "#5C5C5C",
          "fontSize": 14,
          "fontWeight": "400",
          "textTransform": "none",
        },
        undefined,
        {},
      ]
    }
  />
</View>
`;
